homeassistant/components/template/sensor.py

```diff
@@ -54,50 +92,79 @@
 )
 
 
-def trigger_warning(val):
-    """Warn if a trigger is defined."""
+def extra_validation_checks(val):
+    """Run extra validation checks."""
     if CONF_TRIGGER in val:
         raise vol.Invalid(
             "You can only add triggers to template entities if they are defined under `template:`. "
             "See the template documentation for more information: https://www.home-assistant.io/integrations/template/"
         )
 
+    if CONF_SENSORS not in val and SENSOR_DOMAIN not in val:
+        raise vol.Invalid(f"Required key {SENSOR_DOMAIN} not defined")
+
     return val
 
 
+def rewrite_legacy_to_modern_conf(cfg: dict[str, dict]) -> list[dict]:
+    """Rewrite a legacy sensor definitions to modern ones."""
+    sensors = []
+
+    for object_id, entity_cfg in cfg.items():
+        entity_cfg = {**entity_cfg, CONF_OBJECT_ID: object_id}
+
+        for from_key, to_key in LEGACY_FIELDS.items():
+            if from_key not in entity_cfg or to_key in entity_cfg:
+                continue
+
+            val = entity_cfg.pop(from_key)
+            if isinstance(val, str):
+                val = template.Template(val)
+            entity_cfg[to_key] = val
+
+        if CONF_NAME not in entity_cfg:
+            entity_cfg[CONF_NAME] = template.Template(object_id)
+
+        sensors.append(entity_cfg)
+
+    return sensors
+
+
 PLATFORM_SCHEMA = vol.All(
     PLATFORM_SCHEMA.extend(
         {
             vol.Optional(CONF_TRIGGER): cv.match_all,  # to raise custom warning
-            vol.Required(CONF_SENSORS): cv.schema_with_slug_keys(SENSOR_SCHEMA),
+            vol.Optional(CONF_SENSORS): cv.schema_with_slug_keys(LEGACY_SENSOR_SCHEMA),
+            vol.Optional(SENSOR_DOMAIN): vol.All(cv.ensure_list, [SENSOR_SCHEMA]),
```

Comment:
Why should we allow the new format here? Isn't this whole section considered legacy?

Reply:
It was mainly that I had to add support for those entities to accept it _anyway_.

Reply:
but I guess you're right, we should propose to get rid of the platform integration completely. Not deprecate, just don't promote.